include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+    AdditionalData(const SolverType &      solver_type = automatic,
+                   const SolutionStrategy &strategy    = linesearch,
+                   const unsigned int      maximum_non_linear_iterations = 200,
+                   const double            function_tolerance            = 1e-8,
+                   const double            relative_tolerance            = 1e-5,
+                   const double            step_tolerance                = 0.0,
+                   const unsigned int      anderson_subspace_size        = 0);
+
+    /**
+     * The type of nonlinear solver to use. The default value is set to 'auto',
+     * which will use either KINSOL or NOX depending on which package is
+     * installed, KINSOL having priority.
+     */
+    SolverType solver_type;
+
+    /**
+     * The solution strategy to use. For this class, you can choose from
+     * SolutionStrategy::newton, SolutionStrategy::linesearch, or
+     * SolutionStrategy::picard. More details on this can be found on the
+     * @p KINSOL documentation.
+     */
+    SolutionStrategy strategy;
+
+    /**
+     * Maximum number of nonlinear iterations allowed.
+     */
+    unsigned int maximum_non_linear_iterations;
+
+    /**
+     * A scalar used as a stopping tolerance on the scaled
+     * maximum norm of the system function $F(u)$ or $G(u)$.
+     *
+     * If set to zero, default values provided by KINSOL will be used.
+     */
+    double function_tolerance;
+
+    /**
+     * Relative $l_2$ tolerance of the residual to be reached.
+     *
+     * @note Solver terminates successfully if either the absolute or
+     * the relative tolerance has been reached.
+     */
+    const double relative_tolerance;
+
+    /**
+     * A scalar used as a stopping tolerance on the minimum
+     * scaled step length.
+     *
+     * If set to zero, default values provided by KINSOL will be used.
+     */
+    double step_tolerance;
+
+    /**
+     * The size of the subspace used with Anderson acceleration
+     * in conjunction with Picard or fixed-point iteration.
+     *
+     * If you set this to 0, no acceleration is used.
+     */
+    unsigned int anderson_subspace_size;
+  };
+
+  /**
+   * Constructor, filling in default values
+   */
+  NonlinearSolverSelector();
+
+  /**
+   * Constructor, selecting the solver and other parametersspecified in
+   * @p additional_data.
+   */
+  NonlinearSolverSelector(const AdditionalData &additional_data);
+
+  /**
+   * Constructor
+   *
+   * @param additional_data NonlinearSolverSelector configuration data
+   * @param mpi_communicator MPI communicator over which logging operations are
+   * computer.
+   */
+  NonlinearSolverSelector(const AdditionalData &additional_data,
+                          const MPI_Comm &      mpi_communicator);
+
+  /**
+   * Select a new nonlinear solver. All solver names used in this class are
+   * all lower case.
+   */
+  void
+  select(const typename AdditionalData::SolverType &type);
```

Comment:
I don't particularly care for the fact that there are two ways to select a solver: this function and the additional data. Who wins if conflicting information is provided? My preference would be to remove this function.